### DISCUSSION TITLE
Elevation and Apparent Position

### DISCUSSION THREAD

**Discussion by BluesummersC** [skyfielders/python-skyfield#648] [Q&A]
I've been using skyfield to track the Apparent Position of the Sun from a specific lat, long, elevation on Earth. My goal is to track when the sun goes x degrees below the horizon from the view of high elevation.  
 
My confusion is how elevation is used in the calculation, because what I'm seeing from Skyfield does not match what I think I should see. 

At Sea Level, the Apparent Elevation is perfect, and I can easily verify with other 3rd party tools, such as NOAA's sun position calculator.

However, at a high elevation, such as 4km, the Apparent Elevation does not match what I'm seeing. Using skyfield.api.Topos(lat, long, elevation_m), going from 0m to 4000m barely affects the Apparent Elevation of the Sun at all (-2.003456 vs -2.003487), but in reality, standing at 4000m, the Sun takes about 10 more minutes to set, so I'd expect the Apparent Elevation to be a higher value at a higher geographical elevation.

Is my understanding of this flawed? If so, how can I get the Sun's Apparent Elevation at high altitudes?

for reference, my simple code:
 ```
        loader = skyfield.api.Loader(skyfield_dir)
        solar_system = loader(skyfield_bsp)

        # Define our locations
        self.sun = solar_system['sun']
        earth = solar_system['earth']
        our_pos = skyfield.api.Topos(lat, long, elevation_m=4000)
        pos= earth + our_pos
        self.pos= pos

        # grab the current time and get the apparent position of the Sun
        now = skyfield.api.load.timescale(builtin=True).now()
        astrometric = self.pos.at(now).observe(self.sun)
        elevation, azimuth, distance = astrometric.apparent().altaz()

        azimuth = repr(azimuth._degrees)
        elevation = repr(elevation.degrees)
        position = azimuth + ' ' + elevation

        return position
```

**Comment by brandon-rhodes** [comment]
(Alas, this question arrived when I was out of town, and it looks like no other contributors answered it. It might now be too late for your project, but here's my try at an answer:)

* Astronomers always measure elevation (or altitude, as some astronomers call it) from the horizon which is the great circle around the sky exactly 90° between the zenith and the nadir.
* And, if you visit a high altitude location, you'll generally find that the horizon is in the same place because the land around the location is higher too. Denver at a mile high faces an eastern sunrise horizon which is also a mile high, for example.
* If you want to handle the exceptional case of a prominent location like a mountaintop, the documentation explains it here: https://rhodesmill.org/skyfield/almanac.html#elevated-vantage-points
* Let me know if there's a way that section's title or placement can be improved to make it easier for you to find in the future!

**Reply by BluesummersC** [reply]
The link you provided for elevated vantage points was what I was looking for back then. Since you asked, starting at the main page, https://rhodesmill.org/skyfield, I could not find any link or reference to the almanac page. I had no idea that it existed. Even knowing it exists, I went over the Home page and the API Reference page and still couldn't find a way to navigate to the Almanac without typing the URL manually.

My case is a prominent location: Mountain-top to sea-level horizon observation. The calculation works, but I ended up not using it as the light-scatter is still too bright. I ended up just keeping track of what the sun's altitude was at sea-level when it was observably below horizon from the high elevation, and hard-coded that difference as a fudge-factor.

**Reply by brandon-rhodes** [reply]
> I ended up just keeping track of what the sun's altitude was at sea-level when it was observably below horizon from the high elevation, and hard-coded that difference as a fudge-factor.

That's a good approach — the United States Naval Observatory does the same thing for their official almanac, if I recall.

I usually go right from the home page to the documentation Table of Contents, and can Ctrl-F search or else scroll to find the almanac section. Is your problem that you followed another link from the front page on the web, and never found the Table of Contents?

Or is the problem that you found the Table of Contents, but that it didn't display all the subsections of "Risings and Settings", so you couldn't see there was a subsection entitled "Elevated vantage points"?

**Reply by NikosAlexandris** [reply]
Relevant to the original question, how does Skyfield compare to [Pysolar](https://pysolar.readthedocs.io/en/latest/#) when it comes to calculate the solar elevation (altitude) and solar azimuth? Pysolar [documents a validation effort against against code maintained by the United States Naval Observatory (USNO)](https://pysolar.readthedocs.io/en/latest/#validation).

See also: https://pysolar.readthedocs.io/en/latest/#difference-from-pyephem

How should I properly cross-compare the two tools for solar geometry parameters? Mostly I am concerned about the differences in solar azimuth [^1]. Examples:

```
 Longitude   Latitude   Time                        Zone   Model      Declination   Altitude   Azimuth     Units
 ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────
  -71.38200   42.20600   2007-02-18 15:13:01+00:00   UTC    pysolar    NA            30.91426   149.24762   degrees
  -71.38200   42.20600   2007-02-18 15:13:01+00:00   UTC    Skyfield   NA            45.03750   252.66683   degrees
  -71.38200   42.20600   2007-02-18 15:13:01+00:00   UTC    NOAA       -11.86824     30.66639   30.61972    degrees
``` 
[^2]

[^1]: 0 degrees azimuth is due North for all of the above models
[^2]: Own implementation of NOAA's solar geometry equations is mine.

**Reply by brandon-rhodes** [reply]
> How should I properly cross-compare the two tools for solar geometry parameters?

@NikosAlexandris — The traditional approach would be to ask the two libraries about the same location and date, subtract their azimuth angles, and look at the absolute value of the result. (Is that what you were asking?)

**Reply by NikosAlexandris** [reply]
Seems like a hint to the right direction. Does Skyfield document similar efforts like Pysolar?

Note : I am building step-by-step some Pytests.

**Reply by brandon-rhodes** [reply]
> Does Skyfield document similar efforts like Pysolar?

No — writing and maintaining the library is work enough, and I'm happy for third parties and fans to do the work of any comparisons that need to be made. I seem to recall that the AstroPy folks maintain test suites comparing different libraries like Skyfield; you could check whether any of their results are published where users can benefit from them?

**Reply by NikosAlexandris** [reply]
I hope to complete some meaningful tests then, which I can contribute back here and there.
